### ISSUE TITLE
Document review

### ISSUE THREAD

**Issue by ConstanceBeguier** [zkpstandard/wg-plonkish#29]
Hello,

I have some feedback on your technical document.

**Plonkish Relation Section:**
- I found it quite challenging to read the table on instances (especially the last lines) at the beginning of the document because it introduces numerous notations that are explained later in the document.
- To improve the document's readability, I suggest to add simple examples: either a global example that includes each type of constraint or one example per constraint. If it's a global example, it might be beneficial to place it at the beginning of the document (before instance section) to start introducing some notations early on.

**Backend Optimization Section:**
- The purpose of transitioning from an abstract circuit to a concrete circuit is, in my opinion, not clear enough. I wondered if the goal was to optimize the number of rows, the number of columns, or both.
- I have difficulty understanding the last algorithm (algorithm for choosing r). Why don't the hints  h_i appear in it? How can we verify ok_for(...) if the hints h_i are not defined?

Finally, I noticed an inaccuracy in the definition of a multivariate polynomial. I believe that  c_z and nu should not be equal to 0.

**Comment by daira** [comment]
> Finally, I noticed an inaccuracy in the definition of a multivariate polynomial. I believe that $c_z$ and $\nu$ should not be equal to 0.

You are right about $c_z$. Any multivariate polynomial with a $c_z = 0$ term can be rewritten to an equivalent multivariate polynomial with no $c_z = 0$ terms and a smaller $\nu$. However, in order for that to allow expressing the zero polynomial, we must allow $\nu$ to be zero. (We use the common convention that ${\large \Sigma}_{z = 0}^{\nu - 1} (\ldots) = 0$ when $\nu = 0$.)

**Comment by str4d** [comment]
Thanks for the review! @daira, @mmaller and I looked over it today. Comments below are from all of us.

> Plonkish Relation Section:
>
> - I found it quite challenging to read the table on instances (especially the last lines) at the beginning of the document because it introduces numerous notations that are explained later in the document.

Do you mean [this table](https://github.com/zkpstandard/wg-plonkish/blob/main/src/relation.md#instances)?
https://github.com/zkpstandard/wg-plonkish/blob/7f8e78a95bf5edc228069d5df73abd40e88a2a5a/src/relation.md?plain=1#L35-L55

The intention is that this table is defining _variables_, that are specified later in the document. All _notation_ should however have already been either defined or imported by reference here, in the [Dependencies and notation](https://github.com/zkpstandard/wg-plonkish/blob/main/src/relation.md#dependencies-and-notation) section.

From a scan through the table, this is what we currently think might be notation that is not defined at this point:
- $\mathrel{⦂}$ (used to specify types)
- $[0,m) \times [0,n)$ (a product or tuple type)
- $\mathbb{F}^{m \times n}$ (a sequence type)
- $a \mapsto b$ (a mapping that is part of a function or relation)
- $A \rightarrow B$ (a function type) (this will replace the uses of $\mapsto$ in the relation document, which were incorrect)
- Multivariate polynomial

Is there any specific notation (of the above or otherwise) that you particularly found unclear?

> - To improve the document's readability, I suggest to add simple examples: either a global example that includes each type of constraint or one example per constraint. If it's a global example, it might be beneficial to place it at the beginning of the document (before instance section) to start introducing some notations early on.

Part of the issue here is that because the overall documentation is not yet being rendered (because the repo was private for a while), it isn't obvious that the individual pages are not standalone. Once #30 is complete, the rendered version will make clear that `relation.md` and `optimizations.md` are specifications within the overall documentation, and we will consider adding a separate section with examples.

> Backend Optimization Section:
>
> - The purpose of transitioning from an abstract circuit to a concrete circuit is, in my opinion, not clear enough. I wondered if the goal was to optimize the number of rows, the number of columns, or both.

We had rewritten the background in #28 to say this, which I think addresses the comment:
> The benefit of the simpler abstract model is that it allows reasoning about the circuit without needing to be concerned with layout of the witness matrix, i.e. the position of rows relative to each other. The drawback of using only the abstract model, is that it may require a greater number of columns to express a given circuit than a model that supports offsets. This directly affects proof size for popular instantiations of proof systems based on Plonkish.

We've now merged this PR; if you have further feedback on the current language, please let us know.

> - I have difficulty understanding the last algorithm (algorithm for choosing $r$). Why don't the hints $h_i$ appear in it? How can we verify $\mathsf{ok＿for}(...)$ if the hints $h_i$ are not defined?

$\mathsf{ok＿for}(...)$ implicitly closes over both $h_i$ and $e_i$. In particular, if two abstract cells are both constrained, they can't map to the same concrete cell. And the definition of which concrete cell they map to is dependent on the hints.

Would this be clearer if $\mathsf{ok＿for}(...)$ took $\big[ (h_i, e_i) \big]_i$ as an explicit argument?

> Finally, I noticed an inaccuracy in the definition of a multivariate polynomial. I believe that $c_z$ and $\nu$ should not be equal to 0.

You are right about $c_z$. Any multivariate polynomial with a $c_z = 0$ term can be rewritten to an equivalent multivariate polynomial with no $c_z = 0$ terms and a smaller $\nu$. However, in order for that to allow expressing the zero polynomial, we must allow $\nu$ to be zero. (We use the common convention that ${\large \Sigma}_{z = 0}^{\nu - 1} (\ldots) = 0$ when $\nu = 0$.)

**Comment by ConstanceBeguier** [comment]
Thank you for taking into account all my comments on the document.

Regarding your question, you have indeed cited the correct table Instances. When I first read the document, I found it difficult to understand this table because elements like L_v, LOOK_v, and TAB_v are detailed later on. Initially, I thought that to improve the readability of the document, it might be beneficial to introduce the constraints with these notations before introducing the instances and witnesses in details. However, upon further reflection, I am no longer sure if this is the best approach.

**Comment by ConstanceBeguier** [comment]
After reading the new document, I still have a few questions regarding the "Plonkish Backend Optimizations" page:
1. "All constrained abstract cells map to concrete cell coordinates that are in range." What does "in range" mean in this context?
2. "Since correctness does not depend on the specific hints provided by the circuit programmer (...)." I am not sure I fully understand why this is the case.
3. Why are unconstrained abstract cells filled with arbitrary values and fixed abstract cells with 0? Could both types of cells be filled with 0?
4. Regarding the greedy algorithm for choosing r, how are the hints selected/evaluated?

**Comment by ConstanceBeguier** [comment]
This issue has been addressed through PRs #37 , #44, and #47. As these changes cover the scope of the issue, I’m closing it.
